sstables/mx/partition_reversing_data_source.cc

```diff
@@ -0,0 +1,608 @@
+        if (column_value_fix_lengths.empty()) {
+            _ck_column_value_fix_lengths = boost::make_iterator_range(column_value_fix_lengths);
+        } else {
+            _ck_column_value_fix_lengths = boost::make_iterator_range(std::begin(column_value_fix_lengths),
+                                                                      std::begin(column_value_fix_lengths) + _ck_size);
+        }
+        _ck_blocks_header_offset = 0u;
+    }
+    bool no_more_ck_blocks() const { return _ck_column_value_fix_lengths.empty(); }
+    void move_to_next_ck_block() {
+        _ck_column_value_fix_lengths.advance_begin(1);
+        ++_ck_blocks_header_offset;
+        if (_ck_blocks_header_offset == 32u) {
+            _ck_blocks_header_offset = 0u;
+        }
+    }
+    std::optional<uint32_t> get_ck_block_value_length() const {
+        return _ck_column_value_fix_lengths.front();
+    }
+    bool is_block_empty() const {
+        return (_ck_blocks_header & (uint64_t(1) << (2 * _ck_blocks_header_offset))) != 0;
+    }
+    bool is_block_null() const {
+        return (_ck_blocks_header & (uint64_t(1) << (2 * _ck_blocks_header_offset + 1))) != 0;
+    }
+    bool should_read_block_header() const {
+        return _ck_blocks_header_offset == 0u;
+    }
+
+public:
+    bool non_consuming() const {
+        return false;
+    }
+    void verify_end_state() const {
+        if (!_finished) {
+            throw std::runtime_error("row_body_skipping_context - no more data but parsing is incomplete");
+        }
+    }
+    bool end_of_partition() const {
+        return _end_of_partition;
+    }
+    uint64_t prev_len() {
+        return _prev_unfiltered_size;
+    }
+    std::optional<tombstone_reversing_info> current_tombstone_reversing_info() {
+        // std::nullopt if the last consumed unfiltered was not a tombstone
+        return _current_tombstone_reversing_info;
+    }
+    data_consumer::processing_result process_state(temporary_buffer<char>& data) {
+        _processing_data = &data;
+        auto ret = _gen.generate();
+        return ret;
+    }
+private:
+    uint64_t current_position() {
+        return position() - _processing_data->size();
+    }
+    processing_result_generator do_process_state() {
+        while (true) {
+            _finished = false;
+            co_yield read_8(*_processing_data);
+            auto flags = unfiltered_flags_m(_u8);
+            _current_tombstone_reversing_info.reset();
+            if (flags.is_end_of_partition()) {
+                _end_of_partition = true;
+                _finished = true;
+                co_yield data_consumer::proceed::no;
+                break;
+            } else if (flags.is_range_tombstone()) {
+                _current_tombstone_reversing_info.emplace();
+                _current_tombstone_reversing_info->kind_offset = current_position();
+                co_yield read_8(*_processing_data);
```

Comment:
Seems inefficient to yield after each byte unconditionally. Performance is not important for reverse reads, so fine. But is there a reason for this?

Reply:
Because of how coroutines work in c++, we were able to add a condition to the `processing_result_generator`, so that the execution of the coroutine is not suspended despite using `co_yield` keyword (`struct read_awaiter` in processing_result_generator.hh)

Reply:
Right